core/src/main/java/org/apache/accumulo/core/client/lexicoder/BigDecimalLexicoder.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.client.lexicoder;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+import org.apache.accumulo.core.clientImpl.lexicoder.FixedByteArrayOutputStream;
+
+/**
+ * A lexicoder to encode/decode a BigDecimal to/from bytes that maintain its native Java sort order.
+ *
+ * @since 2.1.0
+ */
+public class BigDecimalLexicoder extends AbstractLexicoder<BigDecimal> {
+
+  private final BigIntegerLexicoder bigIntegerLexicoder = new BigIntegerLexicoder();
+
+  @Override
+  public BigDecimal decode(byte[] b) {
+    // This concrete implementation is provided for binary compatibility, since the corresponding
+    // superclass method has type-erased return type Object. See ACCUMULO-3789 and #1285.
+    return super.decode(b);
+  }
+
+  @Override
+  public byte[] encode(BigDecimal bd) {
+    // To encode we separate out the scale and the unscaled value
+    // serialize each value individually and store them
+    int scale = bd.scale();
+    BigInteger bigInt = bd.unscaledValue();
+    byte[] encodedbigInt = bigIntegerLexicoder.encode(bigInt);
+    // Length is set to size of encoded BigInteger + length of the scale value
+    byte[] ret = new byte[4 + encodedbigInt.length];
+    try (DataOutputStream dos = new DataOutputStream(new FixedByteArrayOutputStream(ret))) {
+      scale = scale ^ 0x80000000;
+      dos.write(encodedbigInt);
+      dos.writeInt(scale);
```

Comment:
Wondering if the scale should be added first so that its sorted on first. Thinking of numbers like `4.2E-100`,  `4.2E-10`, `4.2E+10`, `4.2E+100`  

```suggestion
      dos.writeInt(scale);
      dos.write(encodedbigInt);
```

Reply:
Also encoding by adding scale first won't have any bearing. When we sort the encoded value it's just a bunch of bytes. We will have to write our comparator which does what we want.  I'm evaluating few other options, will keep posted.

Reply:
I was thinking the scale was equivalent to an exponent.  The DobuleLexicoder encodes the exponent and then mantissa for proper sorting.  If the scale is not the exponent, then to get proper sorting would need something like an exponent encoded first using a fixed width followed by the mantissa.

Reply:
Here's a good example to show that scale is not equivalent to exponent (paste into `jshell PRINTING` to run):
```java
var y = new BigDecimal("210");
IntStream.rangeClosed(-1, 2).forEach(i -> {
  var x = y.setScale(i);
  printf("unscaledValue() = %s; scale = %s; precision = %s; toString = %s%n", x.unscaledValue(), x.scale(), x.precision(), x);
});
```

Since BigDecimal has arbitrary precision, scale is used to track how the unscaled value must be scaled (how many places to move the decimal) to preserve the intended precision. It's basically the number of `0` digits to add to the number of digits in the unscaled value in order to achieve the desired number of significant digits in the precision.

Reply:
I tried to rewrite this locally to use an exponent and it mostly works.  Below is what I was experimenting with.

```java
public class BigDecimalLexicoder extends AbstractLexicoder<BigDecimal> {

  private final BigIntegerLexicoder bigIntegerLexicoder = new BigIntegerLexicoder();

  @Override
  public BigDecimal decode(byte[] b) {
    // This concrete implementation is provided for binary compatibility, since the corresponding
    // superclass method has type-erased return type Object. See ACCUMULO-3789 and #1285.
    return super.decode(b);
  }

  private static int getExponent(BigDecimal bd){
    return -1*bd.scale()+bd.precision()-1;
  }

  /**
   * inverse of {@link #getExponent(BigDecimal)}
   */
  private static int getScale(int exponent, int precision) {
    return (exponent+1-precision) * -1;
  }

  private byte[] encodeBigInt(BigInteger bigInt) {
    byte[] bytes = bigInt.toByteArray();
    // flip the sign bit
    bytes[0] = (byte) (0x80 ^ bytes[0]);
    return bytes;
  }

  private BigInteger decodeBigInt(byte[] data, int off, int len) {
    // unflip the sign bit
    data[off] = (byte) (0x80 ^ data[off]);
    return new BigInteger(data, off, len);
  }

  @Override
  public byte[] encode(BigDecimal bd) {
    BigInteger bigInt = bd.unscaledValue();
    byte[] encodedbigInt = encodeBigInt(bigInt);
    // Length is set to size of encoded BigInteger + length of the scale value
    byte[] ret = new byte[5 + encodedbigInt.length];
    try (DataOutputStream dos = new DataOutputStream(new FixedByteArrayOutputStream(ret))) {
      // sort first on the sign, make negative numbers come first
      dos.write((byte)bd.signum()  ^ 0x80);

      // sort second on the exponent
      int exponent = getExponent(bd);
      if(bd.signum() < 0) {
        // when the number if negative the exponents need to sort in reverse order
        exponent = exponent * -1;
      }
      dos.writeInt(exponent ^ 0x80000000);

      // sort third on the unscaled value.
      dos.write(encodedbigInt);
      return ret;

    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }

  @Override
  protected BigDecimal decodeUnchecked(byte[] b, int offset, int origLen)
      throws IllegalArgumentException {

    try {
      DataInputStream dis = new DataInputStream(new ByteArrayInputStream(b, offset, origLen));
      int signum = (byte)(dis.read() ^ 0x80);

      int exponent = dis.readInt() ^ 0x80000000;
      if(signum < 0)
        exponent = exponent * -1;

      BigInteger bigInt = decodeBigInt(b,5,b.length - 5);

      // TODO is there a better way to get this
      int precision = bigInt.abs().toString().length();

      return new BigDecimal(bigInt, getScale(exponent, precision));
    } catch (IOException ioe) {
      throw new UncheckedIOException(ioe);
    }
  }
}
```

I modified the test function to use the following test data.


```java
  @Test
  public void testSortOrder() {

    var list = Stream.of("4.2E+10","4.2E+100","4.1E+100","4.3E+100","4.2E-10","4.2E-100",
        "-4.2E+10","-4.2E+100","-4.2E-10","-4.2E-100","-4.1E-100","-4.3E-100","4.2","-4.2","2.1","2.2",
        "2.0", "2.00", "2.000", "-3.000", "-2.00", "0.0000", "0.1", "0.10", "-65537.000", "-65537.123",
        "-65537.00", "-65537.0", "4.1E-10","4.3E-10").map(BigDecimal::new).collect(Collectors.toList());

    assertSortOrder(new BigDecimalLexicoder(), list);

  }
```

And it fails with the following difference in sort order.

```
Expected :[-4.2E+100, -4.2E+10, -65537.123, -65537.000, -65537.00, -65537.0, -4.2, -3.000, -2.00, -4.2E-10, -4.3E-100, -4.2E-100, -4.1E-100, 0.0000, 4.2E-100, 4.1E-10, 4.2E-10, 4.3E-10, 0.1, 0.10, 2.0, 2.00, 2.000, 2.1, 2.2, 4.2, 4.2E+10, 4.1E+100, 4.2E+100, 4.3E+100]

Actual   :[-4.2E+100, -4.2E+10, -65537.00, -65537.0, -65537.123, -65537.000, -4.2, -3.000, -2.00, -4.2E-10, -4.3E-100, -4.2E-100, -4.1E-100, 0.0000, 4.2E-100, 4.1E-10, 4.2E-10, 4.3E-10, 0.1, 0.10, 2.00, 2.000, 2.0, 2.1, 2.2, 4.2, 4.2E+10, 4.1E+100, 4.2E+100, 4.3E+100]
```

So everything in the test data is sorting fine.  The only problem is `-65537.123` is sorting after `-65537.0`.  I understand why its doing this, not sure what to do about it for now though.

Reply:
I came across this paper which I was doing some reading. I think we should to change the encoding to use Bits and hopefully, that would give us what we want.  Here is the link 
[https://arxiv.org/abs/1506.01598]. It has been done in C++ thinking of translating that in Java and see if it gets us what we need.

Reply:
I don't think that paper covers how to deal with the arbitrary precision of BigDecimal. And it's non-trivial to get the exponent from a normalized BigDecimal without explicitly setting the scale, which loses precision.